spec/formatting.md

```diff
@@ -57,6 +57,12 @@ nor be made available to _function handlers_.
 > _declarations_ affecting _variables_ referenced by that _expression_
 > have already been evaluated in the order in which the relevant _declarations_
 > appear in the _message_.
+>
+> Users may write custom functions that have observable side effects.
+> Lazy evaluation may involve evaluating the same expression multiple times
```

Comment:
It's good practice to avoid the normative words like 'may' in non-normative contexts like this.

```suggestion
> Users or implementations can provide functions that have observable side effects.
> Lazy evaluation might involve evaluating the same _expression_ multiple times
```

Notice that I removed the word "custom" (which might be presumed?) and bring in implementations. 

Consider adding a health warning to this paragraph?

Reply:
Removed the "may". What do you mean by "health warning"?

Reply:
By health warning, something like: It's a bad idea to create selectors or formatter functions with such side effects

Reply:
I see. I'm not sure I would agree across the board. For example, you might want a `currentTime` function that returns the system clock time. This doesn't have "side effects" as such, but returns a different value when it's called at different times, so:

```
.local $time = {:currentTime}
{{ {$time} {$time} }}
```

could have a different result depending on whether the implementation evaluates the RHS of the variable `time` once or twice.

(The jargon is that `currentTime` is not referentially transparent.)

Perhaps some language saying that care should be taken when defining such functions and they should be avoided when possible?

Reply:
I thought we already had something in the spec about side effects, but the closest we come to is
> _Function handler_ access to the _formatting context_ MUST be minimal and read-only,
and execution time SHOULD be limited.

which relies on an implementation choosing to call "everything that a function can access" as its _formatting context_, because we've not explicitly prohibited access outside it.

That is to say, I agree with @aphillips that we ought to have stronger language to recommend against functions that mutate anything as a side effect.

Reply:
I added some more language to (a) clarify that the problem also occurs with functions that don't mutate anything, but that depend on external mutable state; and (b) recommend against functions that mutate anything observably.

Reply:
Hmm...

side effect == touches the _resolved value_?

Lots of functions have "side effects" if that's the case. @eemeli your suggestion of limiting the _resolved value_ of `:date` and `:time` would be a side effect if we were to implement it. `currentTime` is a good example. `random` is another. Some operations, such as case-folding, alter the value in an irreversible way. Other operations, such as unfloating a time value, change the meaning.

Our design tries to be immutable, but we're actually mutable to the depth of one expression. Chaining `.local` or reannotating a placeholder can make these side-effects more visible.

Perhaps this paragraphs wants to be more like:

> Evaluating an _expression_ can affect the _resolved value_ associated with 
> a given _variable_.
> When combined with eager vs. lazy evaluation, 
> in which the same _expression_ might be evaluated multiple times,
> the side-effects of the _function_ on the _resolved value_
> might alter the result of formatting the _message_.

Reply:
> side effect == touches the _resolved value_?
>
> Lots of functions have "side effects" if that's the case. @eemeli your suggestion of limiting the _resolved value_ of `:date` and `:time` would be a side effect if we were to implement it.

Defining the _resolved value_ returned by a _function handler_ is not a [side effect](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). Modifying the _operand_ would be a side effect. But a _function handler_ will quite rarely have a reason to return its input _operand_; in general, it will instead create a new _resolved value_ and return that.

> `currentTime` is a good example. `random` is another. Some operations, such as case-folding, alter the value in an irreversible way. Other operations, such as unfloating a time value, change the meaning.

A reasonable _function handler_ for any of these would not have any side effects, but would instead create a new _resolved value_ and return it.

> Our design tries to be immutable, but we're actually mutable to the depth of one expression. Chaining `.local` or reannotating a placeholder can make these side-effects more visible.

Being able to use a _resolved value_ as an _operand_ does not introduce mutability.

Reply:
The phrase "side effect" is clearly confusing, so I rewrote the paragraph without using it. Please take another look!